libbpf-tools/uprobe_helpers.c

```diff
@@ -0,0 +1,263 @@
+{
+	FILE *which;
+	char *cmd, *buf;
+	size_t buf_sz = PATH_MAX;
+
+	if (asprintf(&cmd, "which %s", prog) < 0) {
+		warn("which alloc failed");
+		return NULL;
+	}
+	which = popen(cmd, "r");
+	free(cmd);
+	if (!which) {
+		warn("which failed");
+		return NULL;
+	}
+	buf = malloc(buf_sz);
+	if (!buf) {
+		warn("buf alloc");
+		pclose(which);
+		return NULL;
+	}
+	if (!fgets(buf, buf_sz, which)) {
+		free(buf);
+		pclose(which);
+		return NULL;
+	}
+	/* which has a \n at the end of the string */
+	buf[strlen(buf) - 1] = '\0';
+	pclose(which);
+	return buf;
+}
+
+/*
+ * Returns the full path to the binary for a given pid.  free() the result.
+ * 1) pid == x, binary == ""    : returns the path to x's program
+ * 2) pid == x, binary == "foo" : returns the path to libfoo linked in x
+ * 3) pid == 0, binary == ""    : failure: need a pid or a binary
+ * 4) pid == 0, binary == "bar" : returns the path to `which bar`
+ *
+ * For case 4), ideally we'd like to search for libbar too, but we don't support
+ * that yet.
+ */
+char *get_path_to_binary(const char *binary, pid_t pid)
+{
+	char *path;
+
+	if (!strcmp(binary, "")) {
+		if (!pid) {
+			warn("Uprobes need a pid or a binary\n");
+			return NULL;
+		}
+		return get_path_to_pid(pid);
+	}
+	if (pid)
+		return get_path_to_pidlib(pid, binary);
+
+	/*
+	 * If the user is tracing a program by name, we can find it.  But we
+	 * can't find a library by name yet.  We'd need to parse ld.so.cache or
+	 * something similar.
+	 */
+	path = which_program(binary);
+	if (!path) {
+		warn("Can't find %s (Need a PID if this is a library)\n",
+		     binary);
+	}
+	return path;
+}
+
+/*
+ * Opens an elf at `path` of kind ELF_K_ELF.  Returns NULL on failure.  On
+ * success, close with close_elf(e, fd_close).
+ */
+static Elf *open_elf(const char *path, int *fd_close)
+{
+	int fd;
+	Elf *e;
+
+	if (elf_version(EV_CURRENT) == EV_NONE) {
+		warn("elf init failed\n");
+		return NULL;
+	}
+	fd = open(path, O_RDONLY);
+	if (fd < 0) {
+		warn("Could not open %s\n", path);
+		return NULL;
+	}
+	e = elf_begin(fd, ELF_C_READ, NULL);
+	if (!e) {
+		warn("elf_begin failed: %s\n", elf_errmsg(-1));
+		close(fd);
+		return NULL;
+	}
+	if (elf_kind(e) != ELF_K_ELF) {
+		warn("elf kind %d is not ELF_K_ELF\n", elf_kind(e));
+		elf_end(e);
+		close(fd);
+		return NULL;
+	}
+	*fd_close = fd;
+	return e;
+}
+
+static void close_elf(Elf *e, int fd_close)
+{
+	elf_end(e);
+	close(fd_close);
+}
+
+/* Returns the offset of a function in the elf file `path`, or -1 on failure. */
+off_t func_offset_in_elf(const char *path, const char *func)
+{
+	off_t ret = -1;
+	int fd = -1;
+	Elf *e;
+	Elf_Scn *scn;
+	Elf_Data *data;
+	GElf_Shdr shdr[1];
+	GElf_Sym sym[1];
+	size_t shstrndx;
+
+	e = open_elf(path, &fd);
+
+	if (elf_getshdrstrndx(e, &shstrndx) != 0)
+		goto out;
+
+	scn = NULL;
+	while ((scn = elf_nextscn(e, scn))) {
+		if (!gelf_getshdr(scn, shdr))
+			continue;
+		if (!(shdr->sh_type == SHT_SYMTAB ||
+		      shdr->sh_type == SHT_DYNSYM))
```

Comment:
it hit the 80 column limit.  (82)  though i guess if it was '.' instead of '->', then we'd be ok.  =)

do you all have a limit max?  i usually just run my patches through checkpatch and have vim set up to not go over 80 by default.

Reply:
kernel line length limit was relaxed to 100 characters, so if I can keep the line from wrapping with 100 characters, I usually do